src/osd/ReplicatedBackend.cc

```diff
@@ -236,7 +236,8 @@ void ReplicatedBackend::on_change()
 {
   dout(10) << __func__ << dendl;
   for (auto& op : in_progress_ops) {
-    delete op.second.on_commit;
+    delete op.second->on_commit;
+    op.second->canceled = true;
```

Comment:
instead of introducing the `canceled` field, can we just reset `on_commit` to nullptr. and check it for zero in `ReplicatedBackend::op_commit()`?

Reply:
We could.  I added a new field because it's often valid to have a null finish context, but that turns out not to be the case here.  I think canceled is a bit clearer, but it does use another word of memory.

Reply:
yeah, the reason of a true `cancelded` is different than that of a nullptr `on_commit` here. but do we need to differentiate them for practical purpose or debugging purpose? also, this PR reminds me of #21786. so i think it might be a better (simpler) way to invalidate a finish context by resetting it.